cls/SourceControl/Git/DiscardState.cls

```diff
@@ -0,0 +1,138 @@
+Class SourceControl.Git.DiscardState Extends (%Persistent, %JSON.Adaptor)
+{
+
+Property FullExternalName As %String(MAXLEN = "") [ Required ];
+
+Property Name As %String [ Required ];
+
+Property Contents As %Stream.GlobalCharacter(LOCATION = "^DiscardStateContents");
+
+Property Username As %String [ Required ];
+
+Property Branch As %String [ Required ];
+
+Property Timestamp As %TimeStamp [ Required ];
+
+Property ExternalFile As %Boolean [ Required ];
+
+Method RestoreToFileTree()
+{
+    // Make sure directory for file exists
+    set dir = ##class(%File).GetDirectory(..FullExternalName)
+    if ('##class(%File).DirectoryExists(dir)) {
+        do ##class(%File).CreateDirectoryChain(dir)
+    }
+    
+    // Recreate File
+    set fileStream = ##class(%Stream.FileCharacter).%New()
+    set fileStream.Filename = ..FullExternalName
+    $$$ThrowOnError(fileStream.CopyFrom(..Contents))
+    $$$ThrowOnError(fileStream.%Save())
+
+    // Add file to source-control / IRIS
+    if '..ExternalFile {
+        do ##class(SourceControl.Git.Utils).ImportItem(..Name, 1, 1)
+        do ##class(SourceControl.Git.Utils).AddToServerSideSourceControl(..Name)
+    }
+
+    // Delete discard record
+    $$$ThrowOnError(..%DeleteId(..%Id()))
+}
+
+ClassMethod SaveDiscardState(InternalName As %String, name As %String) As %Status
+{
+        set discardState = ..%New()
+        
+        if (InternalName = "") {
+            // If not in IRIS
+            set externalName = ##class(%File).Construct($Get(^SYS("SourceControl","Git","settings","namespaceTemp")),name)
+            set discardState.FullExternalName = externalName
+            set discardState.Name = name
+            set discardState.ExternalFile = 1
+        } else {
+            set discardState.FullExternalName = ##class(SourceControl.Git.Utils).FullExternalName(InternalName)
+            set discardState.Name = InternalName
+            set discardState.ExternalFile = 0
+        }
+        // Copy over file contents
+        set fileStream = ##class(%Stream.FileCharacter).%New()
+        set fileStream.Filename = discardState.FullExternalName
+        do fileStream.%Open()
+        do discardState.Contents.CopyFrom(fileStream)
+        do fileStream.%Close()
+
+        // Save extra information
+        set discardState.Username = $USERNAME
+        set discardState.Branch = ##class(SourceControl.Git.Utils).GetCurrentBranch()
+        set discardState.Timestamp = $zdatetime($horolog, 3)
+
+        set st = discardState.%Save()
+
+        quit st
+}
+
+ClassMethod DiscardStatesInBranch() As %DynamicArray
+{
+    set currentBranch = ##class(SourceControl.Git.Utils).GetCurrentBranch()
+
+    set query = "SELECT ID FROM SourceControl_Git.DiscardState WHERE branch = ?"
```

Comment:
TODO: filter by repo and branch (in future work, for IPM awareness)

Reply:
Where do we store the repo name?

Reply:
Maybe it would be better to say "context and branch" where context is a namespace or IPM package name.

Reply:
But again, should spin out a separate issue for this.